### PR TITLE
SAM2

### PR DESCRIPTION
**Pull request by hansent**
# Description

Add sam2 to inference. Simplifies the functionality compared to sam. Only adds the ability to cache embeddings/masks, doesn't return embeddings or accept embeddings as input. Also doesn't currently work in the main gpu dockerfile because it needs cuda 12, so we've added another dockerfile. Added a block to the nvidia t4 tests. 

Also, skips `tests/workflows/integration_tests/execution/test_workflow_solving_rock_paper_scissor.py:test_rock_paper_scissors_workflow`, which is 404ing now, maybe someone deleted a necessary resource?

## Type of change

Please delete options that are not relevant.

-   [ ] Bug fix (non-breaking change which fixes an issue)
-   [x] New feature (non-breaking change which adds functionality)
-   [ ] This change requires a documentation update

## How has this change been tested, please provide a testcase or example of how you tested the change?

Add integration tests

## Any specific deployment considerations

New dockerfile, integration tests

## Docs

-   [ ] Docs updated? What were the changes:

### CODE REVIEW

File: docker/dockerfiles/Dockerfile.sam2 [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+    -r requirements.gpu.txt \
+    -r requirements.sdk.http.txt \
+    --upgrade \
+    && rm -rf ~/.cache/pip
+
+# Install setup.py requirements for flash_attn
+RUN python3 -m pip install packaging==24.1  && rm -rf ~/.cache/pip
+
+# Install flash_attn required for Paligemma and Florence2
+# RUN python3 -m pip install -r requirements.pali.flash_attn.txt --no-build-isolation && rm -rf ~/.cache/pip
```
Comment (grzegorz-roboflow) [inline docker/dockerfiles/Dockerfile.sam2:37]: This seems like it's left-over from originating Dockerfile

File: docker/dockerfiles/Dockerfile.sam2 [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+WORKDIR /sam
+RUN apt-get install -y gcc
```
Comment (grzegorz-roboflow) [inline docker/dockerfiles/Dockerfile.sam2:39]: I'd add gcc to the list of packages in first `RUN` command (unless there is good reason to install it separately here)

File: inference/core/roboflow_api.py [hunk 3]
```diff
@@ -225,6 +225,8 @@ def get_roboflow_model_data(
             url=f"{API_BASE_URL}/{endpoint_type.value}/{model_id}",
             params=params,
         )
+        print("URL")
+        print(api_url)
```
Comment (grzegorz-roboflow) [inline inference/core/roboflow_api.py:229]: debug prints

File: inference/models/sam2/segment_anything2.py [hunk 4]
```diff
@@ -0,0 +1,285 @@
...
+                self.embedding_cache[image_id],
+                self.image_size_cache[image_id],
+                image_id,
+            )
+
+        with torch.inference_mode():
+            self.predictor.set_image(img_in)
+            # embedding = self.predictor.get_image_embedding().cpu().numpy()
+            # high_res_feats = [v.cpu().numpy() for v in self.predictor._features["high_res_feats"]]
+            # embedding_dict = {"image_embed": embedding, "high_res_feats": high_res_feats}
```
Comment (grzegorz-roboflow) [inline inference/models/sam2/segment_anything2.py:130]: commented code - should be removed?

File: requirements/requirements.sam2.txt [hunk 5]
```diff
@@ -0,0 +1,4 @@
+rasterio~=1.3
+torch
```
Comment (grzegorz-roboflow) [inline requirements/requirements.sam2.txt:2]: Just a note - should the version of torch be pulled to match CUDA environment?

**Review by grzegorz-roboflow — CHANGES_REQUESTED** [review]

File: requirements/requirements.sam2.txt (hunk 5 above)
Comment (probicheaux) [reply requirements/requirements.sam2.txt:2]: That happens in the dockerfile

File: inference/core/interfaces/http/http_api.py [hunk 6]
```diff
@@ -1310,6 +1329,79 @@ async def sam_segment_image(
...
+                @app.post(
+                    "/sam2/segment_image",
+                    response_model=Sam2SegmentationResponse,
+                    summary="SAM2 Image Segmentation",
+                    description="Run the Meta AI Segmant Anything 2 Model to generate segmenations for image data.",
+                )
+                @with_route_exceptions
+                async def sam2_segment_image(
+                    inference_request: Sam2SegmentationRequest,
+                    request: Request,
```
Comment (yeldarby) [inline inference/core/interfaces/http/http_api.py:1376]: Am I understanding correctly that this doesn't accept boxes or positive/negative points? Only returns all the segmentations for the full image?

File: inference/models/sam2/segment_anything2.py [hunk 7]
```diff
@@ -0,0 +1,282 @@
...
+        """Initializes the SegmentAnything.
+
+        Args:
+            *args: Variable length argument list.
+            **kwargs: Arbitrary keyword arguments.
+        """
+        super().__init__(*args, model_id=model_id, **kwargs)
+        checkpoint = self.cache_file("weights.pt")
+        model_cfg = {
+            "hiera_large": "sam2_hiera_l.yaml",
```
Comment (yeldarby) [inline inference/models/sam2/segment_anything2.py:55]: Did we put these in our weights bucket?

File: inference/core/interfaces/http/http_api.py (hunk 6 above)
Comment (probicheaux) [reply inference/core/interfaces/http/http_api.py:1376]: No, if you look at the [Sam2SegmentationRequest object](https://github.com/roboflow/inference/blob/775094581ef63ec4e2166d4ed6427e93079667aa/inference/core/entities/requests/sam2.py#L59), you will see that it accepts positive/negative points

Comment (probicheaux) [reply inference/core/interfaces/http/http_api.py:1376]: Not accepting boxes right now though

File: inference/models/sam2/segment_anything2.py (hunk 7 above)
Comment (probicheaux) [reply inference/models/sam2/segment_anything2.py:55]: Yes

File: requirements/requirements.sam2.txt (hunk 5 above)
Comment (yeldarby) [reply requirements/requirements.sam2.txt:2]: Is this pip installable also though?

File: inference/models/sam2/segment_anything2.py (hunk 7 above)
Comment (probicheaux) [reply inference/models/sam2/segment_anything2.py:55]: All 4 weights are tested in the nvidia-t4 workflow, a passing run is here: https://github.com/roboflow/inference/actions/runs/10210859531

**Review by yeldarby — APPROVED** [review]

**Pull request merged**